### PR TITLE
New lxd_cpu_effective_total metric

### PR DESCRIPTION
**Pull request by exsver**
Added new metric for containers: lxd_cpus - The total number of CPUs.

### CODE REVIEW

File: lxd/cgroup/abstraction.go [hunk 1]
```diff
@@ -397,6 +397,77 @@ func (cg *CGroup) GetCPUAcctUsage() (int64, error) {
...
+			return -1, err
+		}
+
+		return parseCPUSet(set)
+	}
+
+	return -1, ErrUnknownVersion
+}
+
+// parseCPUSet parse cpuset string and returns the number of CPUs.
```
Comment (stgraber) [inline lxd/cgroup/abstraction.go:425]: `parses a cpuset string`

File: lxd/cgroup/abstraction.go [hunk 2]
```diff
@@ -397,6 +397,77 @@ func (cg *CGroup) GetCPUAcctUsage() (int64, error) {
...
+func parseCPUSet(set string) (int, error) {
+	var out int
+
+	fields := strings.Split(strings.TrimSpace(set), ",")
+	for _, value := range fields {
+		// parse non-range values
```
Comment (stgraber) [inline lxd/cgroup/abstraction.go:431]: `Parse non-range values.`

File: lxd/cgroup/abstraction.go [hunk 3]
```diff
@@ -397,6 +397,77 @@ func (cg *CGroup) GetCPUAcctUsage() (int64, error) {
...
+			_, err := strconv.Atoi(value)
+			if err != nil {
+				return -1, fmt.Errorf("Failed parsing %q: %w", value, err)
+			}
+
+			out++
+			continue
+		}
+
+		// parse ranges
```
Comment (stgraber) [inline lxd/cgroup/abstraction.go:442]: `Parse ranges (should be made of two elements only).`

File: lxd/cgroup/abstraction.go [hunk 4]
```diff
@@ -397,6 +397,77 @@ func (cg *CGroup) GetCPUAcctUsage() (int64, error) {
...
+		valueFields := strings.Split(value, "-")
+		// should be in range format: 1-4 (two elements only)
```
Comment (stgraber) [inline lxd/cgroup/abstraction.go:444]: Drop, covered by comment above.

File: lxd/cgroup/abstraction.go [hunk 5]
```diff
@@ -397,6 +397,77 @@ func (cg *CGroup) GetCPUAcctUsage() (int64, error) {
...
+		if len(valueFields) != 2 {
+			return -1, fmt.Errorf("Failed parsing %q: invalid format", value)
```
Comment (stgraber) [inline lxd/cgroup/abstraction.go:446]: Invalid range format

File: lxd/cgroup/abstraction.go [hunk 6]
```diff
@@ -397,6 +397,77 @@ func (cg *CGroup) GetCPUAcctUsage() (int64, error) {
 	return -1, ErrUnknownVersion
 }
 
+// GetCPUs returns the total number of effective CPUs.
+func (cg *CGroup) GetCPUs() (int, error) {
```
Comment (stgraber) [inline lxd/cgroup/abstraction.go:401]: GetEffectiveCPUs

File: lxd/metrics/types.go [hunk 7]
```diff
@@ -95,6 +97,7 @@ const (
 // MetricNames associates a metric type to its name.
 var MetricNames = map[MetricType]string{
 	CPUSecondsTotal:             "lxd_cpu_seconds_total",
+	CPUs:                        "lxd_cpus",
```
Comment (stgraber) [inline lxd/metrics/types.go:100]: lxd_cpu_effective_total

**Review by stgraber — CHANGES_REQUESTED** [review]
The current commit should be split into:
  - lxd/cgroup: Add GetEffecticeCPUs
  - lxd/metrics: Add lxd_cpu_effective_total

And adding:
 - api: metrics_cpu_effective_total
 
Which contains the usual changes to doc/api-extensions.md and shared/version/api.go

**Comment by simondeziel** [comment]
Considering the `lxd_cpu_seconds_total` already contains a `cpu` label, I'm not sure what that new metrics will be used for?

```
lxd_cpu_seconds_total{cpu="0",mode="system",name="smb",project="default",type="container"} 59
lxd_cpu_seconds_total{cpu="0",mode="user",name="smb",project="default",type="container"} 157
```

**Comment by stgraber** [comment]
@simondeziel that's a good point, you can query the effective CPUs today just by looking at the lxd_cpu_seconds_total count for unique cpu label.

@simondeziel what's the promql query to effectively get the effective CPU count that way?

**Comment by simondeziel** [comment]
@stgraber I think there is a catch with using `lxd_cpu_seconds_total` though. Here's with my instance on your cluster:

```
count(group(lxd_cpu_seconds_total{name="gw-hdc"}) by (cpu))
```

This gives me `32` while there is only 1 CPU core accessible to the instance (as reported by `nproc`). Now if I tell it to ignore CPUs without any time spent:

```
count(group(lxd_cpu_seconds_total{name="gw-hdc"} > 0) by (cpu))
```

I get `8`, presumably because there is no CPU pinning going on.

### Prometheus Node Exporter

It does essentially the same as the first query as we can see in those record rules (https://github.com/prometheus/node_exporter/blob/master/docs/node-mixin/rules/rules.libsonnet):

```
count without (cpu, mode) (
                node_cpu_seconds_total{%(nodeExporterSelector)s,mode="idle"}
              )
```

Translated for LXD we'd get:

```
count without (cpu, mode) (lxd_cpu_seconds_total{name="gw-hdc", mode="system"})
```

Which gives me `32`, again.

@stgraber could you check if my numbers for the `gw-hdc` container match what's in the cpuset file on `abydos`?

**Comment by stgraber** [comment]
```
root@abydos:~# cat /sys/fs/cgroup/cpuset/lxc.payload.sdeziel_gw-hdc/cpuset.effective_cpus 
18
```

So single CPU pinned on that one.

**Comment by simondeziel** [comment]
@stgraber that's odd, I'm now getting result for CPU 18 only but 2 hours ago, the container had touched a bunch of CPUs. Here's what I had:

```
lxd_cpu_seconds_total{name="gw-hdc"} offset 2h > 0
```
|Metric|Value|
|--------|--------|
lxd_cpu_seconds_total{cpu="10", instance="abydos", job="lxd-abydos", mode="system", name="gw-hdc", project="sdeziel", type="container"} | 4
lxd_cpu_seconds_total{cpu="10", instance="abydos", job="lxd-abydos", mode="user", name="gw-hdc", project="sdeziel", type="container"} | 13
lxd_cpu_seconds_total{cpu="13", instance="abydos", job="lxd-abydos", mode="system", name="gw-hdc", project="sdeziel", type="container"} | 9
lxd_cpu_seconds_total{cpu="13", instance="abydos", job="lxd-abydos", mode="user", name="gw-hdc", project="sdeziel", type="container"} | 21
lxd_cpu_seconds_total{cpu="15", instance="abydos", job="lxd-abydos", mode="system", name="gw-hdc", project="sdeziel", type="container"} | 11
lxd_cpu_seconds_total{cpu="15", instance="abydos", job="lxd-abydos", mode="user", name="gw-hdc", project="sdeziel", type="container"} | 23
lxd_cpu_seconds_total{cpu="18", instance="abydos", job="lxd-abydos", mode="system", name="gw-hdc", project="sdeziel", type="container"} | 9
lxd_cpu_seconds_total{cpu="18", instance="abydos", job="lxd-abydos", mode="user", name="gw-hdc", project="sdeziel", type="container"} | 22
lxd_cpu_seconds_total{cpu="20", instance="abydos", job="lxd-abydos", mode="system", name="gw-hdc", project="sdeziel", type="container"} | 10
lxd_cpu_seconds_total{cpu="20", instance="abydos", job="lxd-abydos", mode="user", name="gw-hdc", project="sdeziel", type="container"} | 22
lxd_cpu_seconds_total{cpu="22", instance="abydos", job="lxd-abydos", mode="system", name="gw-hdc", project="sdeziel", type="container"} | 17
lxd_cpu_seconds_total{cpu="22", instance="abydos", job="lxd-abydos", mode="user", name="gw-hdc", project="sdeziel", type="container"} | 39
lxd_cpu_seconds_total{cpu="6", instance="abydos", job="lxd-abydos", mode="system", name="gw-hdc", project="sdeziel", type="container"} | 1
lxd_cpu_seconds_total{cpu="6", instance="abydos", job="lxd-abydos", mode="user", name="gw-hdc", project="sdeziel", type="container"} | 6
lxd_cpu_seconds_total{cpu="7", instance="abydos", job="lxd-abydos", mode="system", name="gw-hdc", project="sdeziel", type="container"} | 3
|lxd_cpu_seconds_total{cpu="7", instance="abydos", job="lxd-abydos", mode="user", name="gw-hdc", project="sdeziel", type="container"} | 7|

File: lxd/cgroup/abstraction.go (hunk 1 above)
Comment (exsver) [reply lxd/cgroup/abstraction.go]: Done

File: lxd/cgroup/abstraction.go (hunk 2 above)
Comment (exsver) [reply lxd/cgroup/abstraction.go]: Done

**Comment by exsver** [comment]
cgroup v2:

<pre>
# cat /sys/fs/cgroup/lxc.payload.ct1003/cpuset.cpus.effective 
0-1
</pre>

<pre>
# lxc query -X GET --wait /1.0/metrics
# HELP lxd_cpu_seconds_total The total number of CPU seconds used in milliseconds.
# TYPE lxd_cpu_seconds_total counter
lxd_cpu_seconds_total{cpu="0",mode="system",name="ct1003",project="default",type="container"} 54
lxd_cpu_seconds_total{cpu="0",mode="user",name="ct1003",project="default",type="container"} 111
# HELP lxd_disk_read_bytes_total The total number of bytes read.
# TYPE lxd_disk_read_bytes_total counter
...
</pre>

<pre>
# lxc exec ct1003 lscpu
Architecture:            x86_64
  CPU op-mode(s):        32-bit, 64-bit
  Address sizes:         40 bits physical, 48 bits virtual
  Byte Order:            Little Endian
CPU(s):                  2
  On-line CPU(s) list:   0,1

</pre>

**Comment by stgraber** [comment]
@simondeziel that's likely because things got re-balanced by LXD.

**Comment by stgraber** [comment]
So based on above (LXD re-balancing), would you agree this change makes sense then?

@simondeziel

**Comment by simondeziel** [comment]
Cgroupv2 doesn't split CPU usage per CPU so the LXD code reports all usage as being on cpu="0" (https://github.com/lxc/lxd/blob/master/lxd/cgroup/abstraction.go#L340), thanks @stgraber for the walk-through.

As such, I think it makes sense to include that new metric. Thanks @exsver for proposing it!

**Pull request merged**